### PR TITLE
Update tests to work with Z3 4.12.1

### PR DESCRIPTION
**Pull request by atomb**
The updates the default version of Z3 to 4.12.1 and updates superficial aspects of the tests (error order, counterexample values, occasional command-line option tweaks) to work with it.

Those test changes have some minor caveats:
* Some of the tests in `DafnyTestGeneration.Test` check very specific things about the solver results that I don't think are required of valid results. I've commented out what I think are the overly-restrictive checks, but they should really be replaced with similar but less restrictive ones. @Dargones, do you have any thoughts?
* Some tests require specific settings of the Z3 parameter `smt.qi.eager_threshold` to succeed, but there isn't a single value that allows them all to succeed. There's even one test I had to split into two files to give two values of this parameter. Let's investigate whether the current value is a good one, in general. Tracked as #3558.
* Dafny is now able to prove an assertion in `Fuel.dfy` that it shouldn't be able to, according to comments in the file. I don't think this is technically either a soundness or incompleteness issue, but it's a strange behavior. Edit: it turns out that the latest nightly of Z3 goes back to being unable to prove that assertion.
* One test (`github-issues/github-issue-2174.dfy`) requires a special setting of the arithmetic solver. This is something we should investigate further on other code bases. Issue #3501 tracks that investigation.
* The `BitvectorsMore` test was missing errors since there were too many in one method. It had `/errorLimit:10`, but that wasn't even high enough. A newer solver version resulted in the order of errors changing, causing known errors to be hidden and previously-untracked errors to be revealed. A higher error limit solves this.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Review by davidcok — DISMISSED** [review]
Just a few changes on the PR itself. Also
- anything that 'needs investigating" should be turned into an Issue (whose number is given in the PR description).
But the big question is this: I'm very wary of requiring users to update to the very latest OS. Can we not support whatever Z3 is needed to go back at least one OS version? And automatically make the recommended options for Z3. As a user, I need to be concerned with proving my program -- I would not want to have to be a Z3 expert (or even know that Z3 is there).  Getting or building 4.12.1 for older OSes is likely the easier course -- but should be taken before we spring this on users.

File: Source/DafnyTestGeneration.Test/Various.cs
Comment (Dargones) [inline Source/DafnyTestGeneration.Test/Various.cs:156]: I think changing this to `methods.Count is >= 2 and <= 6` would make sure the assertion accounts for all possible solutions.

**Comment by Dargones** [comment]
Thank you, @atomb for pinging about test generation tests. The changes make sense to me - I will look over other tests too to see if there are any potential problems.

File: Test/allocated1/dafny0/BitvectorsMore.dfy
Comment (atomb) [reply Test/allocated1/dafny0/BitvectorsMore.dfy:1]: Just to be clear, this _increases_ the error limit beyond the default of 5, so that we collect all errors that happen. I just looked, though, and using `0` would tell it to find as many as possible, so that'd be a better choice here, and everywhere in the test suite.

I generally agree that tests with a large number of errors should be broken up, but the particular case where this was necessary feels, to me, better as a unit than multiple pieces.

File: Test/dafny0/FunctionSpecifications.dfy
Comment (atomb) [reply Test/dafny0/FunctionSpecifications.dfy:52]: Ah, I didn't mean to leave that commented out. Good catch!

File: Source/DafnyTestGeneration.Test/Various.cs
Comment (atomb) [reply Source/DafnyTestGeneration.Test/Various.cs:156]: Thanks!

**Comment by atomb** [comment]
> Thank you, @atomb for pinging about test generation tests. The changes make sense to me - I will look over other tests too to see if there are any potential problems.

Thanks for confirming!

**Comment by atomb** [comment]
> Just a few changes on the PR itself. Also
> 
>     * anything that 'needs investigating" should be turned into an Issue (whose number is given in the PR description).
>       But the big question is this: I'm very wary of requiring users to update to the very latest OS. Can we not support whatever Z3 is needed to go back at least one OS version? And automatically make the recommended options for Z3. As a user, I need to be concerned with proving my program -- I would not want to have to be a Z3 expert (or even know that Z3 is there).  Getting or building 4.12.1 for older OSes is likely the easier course -- but should be taken before we spring this on users.

I wasn't clear in the original description, but most of the "needs investigating" I plan to resolve before merging this. But, for any that remain, I'll definitely create an issue.

As to requiring recent Linux distributions, I totally agree. My open question is how best to resolve it, and I was hoping reviewers may have thoughts. The primary options I see are 1) building it ourselves as part of CI (which would slow down our builds quite a lot), 2) building a binary like the one on the Z3 GitHub releases page but compatible with older Linux distributions and putting it somewhere generally accessible (where?), or 3) waiting a little while to see if the Z3 developers respond to the issue I opened about their Linux binaries (but we don't want to wait very long).

**Comment by atomb** [comment]
Update on the Z3 binary issue: they've already updated the nightly builds to go back to ubuntu-20.04, and are looking into how they could update the release.

File: Test/allocated1/dafny0/BitvectorsMore.dfy
Comment (davidcok) [reply Test/allocated1/dafny0/BitvectorsMore.dfy:1]: Yes - we should make errorLimit=0 the default in the tests for everything -- both in lit and in the CI tests. I'll add that to my Dafny 4 run and see if we catch anything that way.

**Comment by atomb** [comment]
Several updates:
* This now uses the nightly 4.12.2 pre-release build as a test of how well those builds work for us. This causes some tests to fail again. Shortly, I'm going to update this to use the self-built Z3 binaries from https://github.com/dafny-lang/solver-builds, which should resolve those test failures.
* The Windows DafnyLanguageServer.Test runs are running out of memory. I'm trying to diagnose why.
* I think I've addressed all other comments on this so far.

**Comment by atomb** [comment]
The Z3 builds used by this PR are now custom-built from the release commits of Z3 4.12.1, hosted [here](https://github.com/dafny-lang/solver-builds). (Binaries for 4.8.5 are also available in that form.)

The only remaining issue I'm aware of is that of the language server unit tests running out of memory.

**Review by robin-aws — CHANGES_REQUESTED** [review]
Conservatively blocking this until it retargets `main-4.x` instead (and most of the content will be merged in separate PRs first)

File: Test/allocated1/Allocated1.dfy
Comment (atomb) [inline Test/allocated1/Allocated1.dfy:2]: This might make sense to include with 3.x, too. Same with other instances of `/errorLimit:0`.

File: Test/allocated1/dafny0/BitvectorsMore.dfy.expect
Comment (atomb) [inline Test/allocated1/dafny0/BitvectorsMore.dfy.expect:12]: The extra errors are an artifact of `/errorLimit:0`.

File: Test/cli/errorLimit.dfy
Comment (atomb) [inline Test/cli/errorLimit.dfy:1]: This change might make sense to include for 3.x, as well. Any error limit other than 0 can lead to non-determinism.

File: Test/dafny0/Termination.dfy
Comment (atomb) [inline Test/dafny0/Termination.dfy:490]: I discovered these during the creation of this branch, but it couldn't hurt to include them in 3.x, as well.

File: Test/dafny2/SnapshotableTrees.dfy
Comment (atomb) [inline Test/dafny2/SnapshotableTrees.dfy:1]: I should probably switch this back to 4 and add back in the execution output.

File: Test/dafny4/SoftwareFoundations-Basics.dfy
Comment (atomb) [inline Test/dafny4/SoftwareFoundations-Basics.dfy:272]: This could also be included for 3.x

**Comment by atomb** [comment]
The things that should be included in 3.x are now in #3628. We can merge that into`master` and update `main-4.0` to include it, as well.

File: Test/dafny0/BitvectorsMore.dfy.expect
Comment (atomb) [inline Test/dafny0/BitvectorsMore.dfy.expect:486]: The larger number of errors here is due to `/errorLimit:0`.

File: .github/workflows/doc-tests.yml
Comment (robin-aws) [inline .github/workflows/doc-tests.yml:39]: FYI (to fix before the next time we change the version number :), it's not too hard to extract out the value from source code to avoid duplicating it all over the place in these files. For example: https://github.com/dafny-lang/dafny/blob/master/.github/workflows/msbuild.yml#L35

File: Source/DafnyLanguageServer.Test/Lookup/HoverVerificationTest.cs
Comment (robin-aws) [inline Source/DafnyLanguageServer.Test/Lookup/HoverVerificationTest.cs:69]: Do you know why we lost this?

File: Test/allocated1/dafny0/Computations.dfy
Comment (robin-aws) [inline Test/allocated1/dafny0/Computations.dfy:1]: Is this an option we should (right now) suggest as a verification technique in the manual and (slightly later) add a Dafny-level option for of some kind?

File: Test/dafny0/AutoContracts.dfy.expect
Comment (robin-aws) [inline Test/dafny0/AutoContracts.dfy.expect:589]: Not related to this PR, but why do we have "N1" as a line number?

File: Test/dafny2/SnapshotableTrees.dfy
Comment (robin-aws) [reply Test/dafny2/SnapshotableTrees.dfy:1]: Yup :)

File: Test/git-issues/git-issue-2301.dfy
Comment (robin-aws) [inline Test/git-issues/git-issue-2301.dfy:51]: Why did this have to change?

File: Test/vstte2012/BreadthFirstSearch.dfy
Comment (robin-aws) [inline Test/vstte2012/BreadthFirstSearch.dfy:115]: Why is it okay to remove these?

**Review by robin-aws — APPROVED** [review]
WHEW. I do want to discuss my questions but merging this to get a RC for Dafny 4.0 earlier. I cut #3651 to make sure we come back to them before releasing.

File: .github/workflows/doc-tests.yml
Comment (atomb) [reply .github/workflows/doc-tests.yml:39]: Yeah, that would be a nice improvement.

File: Source/DafnyLanguageServer.Test/Lookup/HoverVerificationTest.cs
Comment (atomb) [reply Source/DafnyLanguageServer.Test/Lookup/HoverVerificationTest.cs:69]: The `???` matches anything, so that change just makes the test less fragile as the solver evolves.

File: Test/allocated1/dafny0/Computations.dfy
Comment (atomb) [reply Test/allocated1/dafny0/Computations.dfy:1]: I've considered that. I'd like to experiment with it a bit more on production code and see if it makes any difference there.

File: Test/dafny0/AutoContracts.dfy.expect
Comment (atomb) [reply Test/dafny0/AutoContracts.dfy.expect:589]: No idea! :)

File: Test/vstte2012/BreadthFirstSearch.dfy
Comment (atomb) [reply Test/vstte2012/BreadthFirstSearch.dfy:115]: They were intermediate assertions introduced to make it easier to prove the loop invariant. But it turned out that only one of them was necessary. The rest just made Z3 work harder.

Comment (robin-aws) [reply Test/vstte2012/BreadthFirstSearch.dfy:115]: Thanks, I suspected as much. It's hard to tell on a quick glance which is the "important" assertion. Perhaps specification coverage mechanisms would help here. :)

File: Test/dafny0/AutoContracts.dfy.expect
Comment (robin-aws) [reply Test/dafny0/AutoContracts.dfy.expect:589]: Oh I actually missed that it was in addition to the actual source location. It's the first time I've encountered that.

**Pull request merged**
